### PR TITLE
[Merged by Bors] - fix: multigoal linter should not flag tactics that do nothing

### PR DESCRIPTION
**Pull request by kmill**
When a tactic does not change the list of goals, the multigoal linter was giving incorrect error messages such as
```
There are 2 unclosed goals before 'change ...' and
at least one remaining goal afterwards.
Please focus on the current goal, for instance
using `·` (typed as "\.").
```
This PR adjusts the condition to allow all tactics that operate on one goal and result in zero or more goals. It also makes the error message more precise, for example
```
warning: The following tactic starts with 2 goals and
ends with 1 goal, 1 of which is not operated on.
  exact .intro
Please focus on the current goal, for instance
using `·` (typed as "\.").
```

---

[Open in Gitpod](https://gitpod.io/from-referrer/)

### CODE REVIEW

File: Mathlib/Tactic/Linter/Multigoal.lean [hunk 1]
```diff
@@ -128,21 +128,25 @@ abbrev ignoreBranch : Std.HashSet SyntaxNodeKind := .ofList [
 
 /-- `getManyGoals t` returns the syntax nodes of the `InfoTree` `t` corresponding to tactic calls
 which
-* leave at least one goal that was not present before it ran;
+* leave at least one goal that was present before it ran
+  (with the exception of tactics that leave the sole goal unchanged);
 * are not excluded through `exclusions` or `ignoreBranch`;
 
 together with the total number of goals.
```
Comment (adomani) [inline Mathlib/Tactic/Linter/Multigoal.lean:135]:
```suggestion
 together with the number of goals before the tactic,
the number of goals after the tactic, and the number of unaffected goals.
```

**Review by adomani — APPROVED** [review]

**Comment by adomani** [comment]
This looks great to me, thanks!

I wonder whether some of the exceptions can now be removed: maybe adaptation_note, sleep_heartbeats, subst_vars?  I have not tested these, though!

**Review by jcommelin — COMMENTED** [review]
bors d=adomani

**Comment by adomani** [comment]
Kyle, while looking at the test, I realized that it still had the "old" workaround of setting the nested linter options and for this reason, the test that you added was not actually linted.  I removed the workaround and now everything should be actually linted more directly!

**Comment by adomani** [comment]
bors merge

**Pull request closed without merging**